ce/api/util.py

```diff
@@ -13,33 +13,35 @@
 from ce.api.geo import wkt_to_masked_array
 
 
-def get_files_from_run_variable(run, variable):
-    return [
-        file_
-        for file_ in run.files
-        if variable in [dfv.netcdf_variable_name for dfv in file_.data_file_variables]
-    ]
+def get_files_from_run_variable(run, variable, ensemble_name):
+    files = []
+    for file_ in run.files:
+        for dfv in file_.data_file_variables:
+            for ensemble in dfv.ensembles:
+                if (
+                    variable == dfv.netcdf_variable_name
+                    and ensemble.name == ensemble_name
+                ):
+                    files.append(file_)
+    return files
```

Comment:
There is a tighter and more efficient way of accomplishing this:

```python
return list(filter(
    lambda file_: any(
        variable == dfv.netcdf_variable_name and 
            any(ensemble.name == ensemble_name for ensemble in dfv.ensembles)
        for dfv in file_.data_file_variables
    ),
    run.files
))
```

Generally speaking, as in R, loops are a code smell, or at least a reason to look at alternatives. Almost always there is a comprehension or a `filter` expression that will do the same job as a something that builds up a result by weeding through another iterable.

You don't have to wrap the `filter` in `list` unless the result must be a list instead of a generic iterable.

Not sure if Black will agree with my spacing there.

Reply:
There's an alternative that uses a comprehension instead of a filter. It may be easier to read:

```python
return (
    file_ for file_ in run.files if any(...)
)
```

The parentheses (`return (...)`) make this a generator. If you used brackets (`return [...]`) it would be a list.

Reply:
These are pretty much identical formulations.

Reply:
Thank you for the advice! I have just been working on this when my other projects are slow.